### PR TITLE
CORTX-26259:CSM: Support Config SW Upgrade

### PR DESCRIPTION
**Pull request by Pranali04796**
Signed-off-by: Pranali04796 <pranali.ugale@seagate.com>

# Problem Statement
- Support Config SW Upgrade for csm
- JIRA ID -  [CORTX-26259](https://jts.seagate.com/browse/CORTX-26259) 

# Design
-  User should upgrade image of deployment without ambiguity.
-   [Design document](https://seagate-systems.atlassian.net/wiki/spaces/PUB/pages/785448990/CORTX+Software+Upgrade)
- Command for upgrade (CSM)-
- -  `csm_setup upgrade --config yaml:///etc/cortx/cluster.conf --services agent`

# Coding
   Checklist for Author
-  [ ] Coding conventions are followed and code is consistent

# Testing 
  Checklist for Author
- [x] Unit and System Tests are added
- [x] Test Cases cover Happy Path, Non-Happy Path and Scalability
- [x] Testing was performed with RPM
- Build Details:
m1
- Deployment Details:
m5
Upgrade Command:
m4
- Test case scenarios tested on above build
- https://seagate-systems.atlassian.net/wiki/spaces/~340452269/pages/949945098/Software+upgrade+-+CSM

# Impact Analysis
  Checklist for Author/Reviewer/GateKeeper
- [ ] Interface change (if any) are documented
- [ ] Side effects on other features (deployment/upgrade)
- [ ] Dependencies on other component(s)

# Review Checklist 
  Checklist for Author
- [x] JIRA number/GitHub Issue added to PR
- [x] PR is self reviewed
- [x] Jira and state/status is updated and JIRA is updated with PR link
- [ ] Check if the description is clear and explained

# Documentation
  Checklist for Author
- [ ] Changes done to WIKI / Confluence page / Quick Start Guide

### CODE REVIEW

File: csm/conf/upgrade.py [hunk 1]
```diff
@@ -28,5 +30,114 @@ def __init__(self):
...
+            Log.info("Loading default csm configurations.")
+            self.load_default_config()
+        except KvError as e:
+            Log.error(f"Configuration Loading Failed {e}")
+            raise CsmSetupError("Could Not Load Url Provided in Kv Store.")
+
+        services = command.options.get("services")
+        if ',' in services:
+            services = services.split(",")
+        elif 'all' in services:
```
Comment (pawankumarsrivastava) [inline csm/conf/upgrade.py:49]: This is not required. Web and CLI are not part of CSM Agent now. You should throw error if other than all or agent is given.
All = agent only

File: csm/conf/upgrade.py [hunk 2]
```diff
@@ -28,5 +30,114 @@ def __init__(self):
 
     async def execute(self, command):
         Log.info("Perform upgrade for csm_setup")
-        # TODO: Implement upgrade logic
+        try:
+            Log.info("Loading Url into conf store.")
```
Comment (pawankumarsrivastava) [inline csm/conf/upgrade.py:34]: 3 Log entries are not required. You could say that Performing upgrade and loading config files

File: csm/conf/upgrade.py [hunk 3]
```diff
@@ -28,5 +30,114 @@ def __init__(self):
...
+            raise CsmSetupError("Could Not Load Url Provided in Kv Store.")
+
+        services = command.options.get("services")
+        if ',' in services:
+            services = services.split(",")
+        elif 'all' in services:
+            services = ["agent", "web", "cli"]
+        else:
+            services=[services]
+        self.execute_web_and_cli(command.options.get("config_url"),
```
Comment (pawankumarsrivastava) [inline csm/conf/upgrade.py:53]: This is not required

File: csm/conf/upgrade.py [hunk 4]
```diff
@@ -28,5 +30,114 @@ def __init__(self):
...
+        Log.info("Preparing for upgrade.")
+        self._update_general_config(const.CSM_DEFAULT_CONF_INDEX, const.CSM_GLOBAL_INDEX)
+        self._update_db_config(const.CSM_DEFAULT_DB_CONF_INDEX, const.DATABASE_INDEX)
+
+    def _update_general_config(self, default_index, current_index):
+        Log.info("Updating general configurations.")
+        self._update(default_index, current_index)
+
+    def _update_db_config(self, default_index, current_index):
+        Log.info("Updating database related configurations.")
```
Comment (pawankumarsrivastava) [inline csm/conf/upgrade.py:85]: Updating database related configurations -> Updating database configurations.

File: csm/conf/upgrade.py [hunk 5]
```diff
@@ -28,5 +30,114 @@ def __init__(self):
...
+        :param current_index:current configutions of deployed version.
+        :returns
+        """
+        default_keys =  Conf.get_keys(default_index)
+        for key in default_keys:
+            default_value = Conf.get(default_index, key)
+            # default_val is empty i,e expecting value from conf_store
+            if not default_value:
+                continue
+
```
Comment (pawankumarsrivastava) [inline csm/conf/upgrade.py:101]: Remove empty line

File: csm/conf/upgrade.py [hunk 6]
```diff
@@ -28,5 +30,114 @@ def __init__(self):
...
+    def _previous_key(self, key):
+        """
+        Form previous key from given key.
+        if key is 'DEBUG>enabled'
+        previous_key will be 'DEBUG>pre_enabled'
+        """
+        keys = key.rsplit('>', 1)
+        if len(list)==1:
+            return f"pre_{key}"
+        return keys[0]+">"+f"pre_{keys[-1]}"
```
Comment (pawankumarsrivastava) [inline csm/conf/upgrade.py:143]: Use fsrting instead of '+'

**Review by pawankumarsrivastava — CHANGES_REQUESTED** [review]

**Comment by pawankumarsrivastava** [comment]
Please add testing results and build information

File: csm/conf/upgrade.py (hunk 1 above)
Comment (Pranali04796) [reply csm/conf/upgrade.py]: Resolved.

File: csm/conf/upgrade.py (hunk 2 above)
Comment (Pranali04796) [reply csm/conf/upgrade.py]: Removed extra log entries

File: csm/conf/upgrade.py (hunk 3 above)
Comment (Pranali04796) [reply csm/conf/upgrade.py]: Removed web and cli part

File: csm/conf/upgrade.py (hunk 4 above)
Comment (Pranali04796) [reply csm/conf/upgrade.py]: Addressed

File: csm/conf/upgrade.py (hunk 5 above)
Comment (Pranali04796) [reply csm/conf/upgrade.py]: Removed empty lines

File: csm/conf/upgrade.py (hunk 6 above)
Comment (Pranali04796) [reply csm/conf/upgrade.py]: Updated using fsrting insted on '+'

File: csm/conf/upgrade.py [hunk 7]
```diff
@@ -27,6 +29,90 @@ def __init__(self):
...
-        Log.info("Perform upgrade for csm_setup")
-        # TODO: Implement upgrade logic
+        Log.info("Performing upgrade and loading config files")
+        try:
+            Conf.load(const.CONSUMER_INDEX, command.options.get(const.CONFIG_URL))
+            self.load_csm_config_indices()
+            self.load_default_config()
+        except KvError as e:
+            Log.error(f"Configuration Loading Failed {e}")
+            raise CsmSetupError("Could Not Load Url Provided in Kv Store.")
```
Comment (pawankumarsrivastava) [inline csm/conf/upgrade.py:39]: Could Not Load Url Provided in Kv Store -> Unable to load configurations

File: csm/conf/configure.py [hunk 8]
```diff
@@ -136,6 +136,8 @@ def set_csm_endpoint(self):
         # Conf.set(const.CSM_GLOBAL_INDEX, const.AGENT_HOST, csm_host)
         if csm_protocol == 'https':
             Conf.set(const.CSM_GLOBAL_INDEX, const.CSM_DEBUG_MODE, 'false')
+        else:
+            Conf.set(const.CSM_GLOBAL_INDEX, const.CSM_DEBUG_MODE, 'true')
```
Comment (pawankumarsrivastava) [inline csm/conf/configure.py:140]: Why is this required?

**Review by pawankumarsrivastava — CHANGES_REQUESTED** [review]

File: csm/conf/configure.py (hunk 8 above)
Comment (Pranali04796) [reply csm/conf/configure.py:140]: We are considering empty default value for those keys whose value is depend on conf store. 
Here i can see we are adding "true" only based on value of conf store.  we must provide value for each key if it depends on conf store.

File: csm/conf/upgrade.py (hunk 7 above)
Comment (Pranali04796) [reply csm/conf/upgrade.py]: Resolved

File: csm/conf/configure.py (hunk 8 above)
Comment (pawankumarsrivastava) [reply csm/conf/configure.py:140]: The debug mode config is not required for CSM now. We can take this change inthe next sprint. 
For now it is fine.

**Review by pawankumarsrivastava — APPROVED** [review]

**Review by rohitkolapkar — APPROVED** [review]
